Cpp_hands_on_programming/lib/abstract/Ipc_method.cc

```diff
@@ -0,0 +1,60 @@
+#include "Ipc_method.h"
+#include <iostream>
+
+int Ipc_method::getFileSize(std::string filename) {
+    int size;
+    std::ifstream in(filename, std::ios::binary);
+	if (!in) {
+		perror("getFileSize");
+		exit(EXIT_FAILURE);
+	}
+	in.seekg(0, std::ios::end);
+	size = in.tellg();
+    in.close();
+    return size;
+}
+
+int Ipc_method::readFile(std::string filename, int offset, std::vector<char> &buffer, int size) {
+    std::ifstream in(filename, std::ios::binary);
+	if (!in.is_open()) {
+		perror("ifstream::open\n");
+		return -1;
+	}
+	// move file position according to offset
+	in.seekg(offset, std::ios::beg);
+	if (!in) {
```

Comment:
This one.

Reply:
From my understanding, seekg return the istream object it is associated with, if it fail the internal error flag of istream will be set, checking that state of `in` can determine that

https://www.cplusplus.com/reference/istream/istream/seekg/

Reply:
rdstate is good. I think it just depends on your preference in this case. Basically we can use ios::fail() to determine the error state(fail() the same as at least failbit or badbit is true). And `fail()` is the synonym operator  `!`. 

There is no problem using `!in` to check the error state of the stream, because it is the same as checking failbit and badbit.

Reply:
Ok I found some explications for this behavior: https://en.cppreference.com/w/cpp/io/basic_ios/operator_bool

Sorry for bothering you.

Reply:
Not at all xD. Explaining help me understand things on a deeper level